modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,225 @@
+    setInput(impl_->blob, "", impl_->scale, impl_->mean);
+
+    if (getLayer(0)->outputNameToIndex("im_info") != -1) {  // Faster-RCNN or R-FCN
+        frameWidth  = impl_->size.width;
+        frameHeight = impl_->size.height;
+        Mat imInfo = (Mat_<float>(1, 3) << impl_->size.height, impl_->size.width, 1.6f);
+        setInput(imInfo, "im_info");
+    }
+
+    boxes.clear();
+    confidences.clear();
+    classIds.clear();
+
+    std::vector<Mat> detections;
+    impl_->outNames = getUnconnectedOutLayersNames();
+    forward(detections, impl_->outNames);
+    CV_Assert(!detections.empty());
+
+    std::vector<String> layerNames = getLayerNames();
+    int lastLayerId = getLayerId(layerNames.back());
+    Ptr<Layer> lastLayer = getLayer(lastLayerId);
+
+    std::vector<int> predClassIds;
+    std::vector<Rect2d> predBoxes;
+    std::vector<float> predConf;
+
+    if (lastLayer->type == "DetectionOutput") {
+        // Network produces output blob with a shape 1x1xNx7 where N is a number of
+        // detections and an every detection is a vector of values
+        // [batchId, classId, confidence, left, top, right, bottom]
+        for (int i = 0; i < detections.size(); ++i)
+        {
+            float* data = (float*)detections[i].data;
+            for (int j = 0; j < detections[i].total(); j += 7)
+            {
+                float conf = data[j + 2];
+                if (conf < confThreshold)
+                    continue;
+
+                float left   = data[j + 3];
+                float top    = data[j + 4];
+                float right  = data[j + 5];
+                float bottom = data[j + 6];
+                float width  = right  - left + 1;
+                float height = bottom - top  + 1;
+
+                if (absoluteCoords && width * height <= 1)
+                    boxes.emplace_back(left * frameWidth, top * frameHeight,
+                                       width * frameWidth, height * frameHeight);
+                else
+                    boxes.emplace_back(left, top, width, height);
+
+                classIds.push_back(static_cast<int>(data[j + 1]));
+                confidences.push_back(conf);
+            }
+        }
+    } else if (lastLayer->type == "Region") {
+        for (int i = 0; i < detections.size(); ++i) {
+            for (int j = 0; j < detections[i].rows; ++j) {
+
+                Mat scores = detections[i].row(j).colRange(5, detections[i].cols);
+                Point classIdPoint;
+                double conf;
+                minMaxLoc(scores, nullptr, &conf, nullptr, &classIdPoint);
+
+                if (conf < confThreshold)
+                    continue;
+
+                std::vector<Range> coord = {Range(j, j + 1), Range(0, 5)};
+                Mat coords;
+                detections[i](coord).copyTo(coords);
+
+                std::vector<float> bboxes = coords.reshape(1, 1);
```

Comment:
Please refactor "Region" output layer postprocessing as well. Use `float*` instead of vectors.

Reply:
done